netebpfext/user/kernel_um.cpp

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include <condition_variable>
+#include <mutex>
+
+#include "netebpfext_platform.h"
+
+#include "kernel_um.h"
+
+typedef struct _ebpf_rundown_ref
+{
+    std::mutex lock;
+    std::condition_variable cv;
+    size_t count = 0;
+    bool rundown_in_progress = false;
+} ebpf_rundown_ref;
+
+typedef struct _MDL
+{
+    MDL* next;
+    size_t size;
+    uint64_t flags;
+    void* start_va;
+    unsigned long byte_offset;
+    unsigned long byte_count;
+} MDL, *PMDL;
+
+typedef struct _IO_WORKITEM
+{
+    DEVICE_OBJECT* device;
+    PTP_WORK work_item;
+    IO_WORKITEM_ROUTINE* routine;
+    void* context;
+} IO_WORKITEM;
+
+typedef ULONG PFN_NUMBER;
+#define PAGE_SIZE 4096
+#define PAGE_SHIFT 12L
+
+#define PAGE_ALIGN(Va) ((PVOID)((ULONG_PTR)(Va) & ~(PAGE_SIZE - 1)))
+#define BYTE_OFFSET(Va) ((ULONG)((LONG_PTR)(Va) & (PAGE_SIZE - 1)))
+#define ADDRESS_AND_SIZE_TO_SPAN_PAGES(Va, size)                                                        \
+    (((((size)-1) >> PAGE_SHIFT) +                                                                      \
+      (((((ULONG)(size - 1) & (PAGE_SIZE - 1)) + (PtrToUlong(Va) & (PAGE_SIZE - 1)))) >> PAGE_SHIFT)) + \
+     1L)
+
+unsigned long
+MmGetMdlByteCount(_In_ MDL* mdl)
+{
+    return mdl->byte_count;
+}
+
+#define MmGetMdlByteOffset(mdl) ((mdl)->byte_offset)
+#define MmGetMdlBaseVa(mdl) ((mdl)->start_va)
+#define MmGetMdlVirtualAddress(mdl) ((PVOID)((PCHAR)((mdl)->start_va) + (mdl)->byte_offset))
+#define MmInitializeMdl(mdl, base_va, length)                                                                     \
+    {                                                                                                             \
+        (mdl)->next = (PMDL)NULL;                                                                                 \
+        (mdl)->size =                                                                                             \
+            (uint16_t)(sizeof(MDL) + (sizeof(PFN_NUMBER) * ADDRESS_AND_SIZE_TO_SPAN_PAGES((base_va), (length)))); \
+        (mdl)->flags = 0;                                                                                         \
+        (mdl)->start_va = (PVOID)PAGE_ALIGN((base_va));                                                           \
+        (mdl)->byte_offset = BYTE_OFFSET((base_va));                                                              \
+        (mdl)->byte_count = (ULONG)(length);                                                                      \
+    }
+
+unsigned long __cdecl DbgPrintEx(
```

Comment:
This is a relic from the early prototype days. We may want to replace this with Tracelog.

Reply:
It's used by KdPrintEx. I propose we keep the switch from KdPrintEx -> tracelog as a separate PR.

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/1466